packages/block-library/src/navigation/index.php

```diff
@@ -121,6 +121,14 @@ function render_block_core_navigation( $attributes, $content, $block ) {
 
 	unset( $attributes['rgbTextColor'], $attributes['rgbBackgroundColor'] );
 
+	wp_enqueue_script(
+		'core_block_navigation_load_modals',
+		plugins_url( 'gutenberg/build/modals/index.js' ),
```

Comment:
I'd love for this to be something like `frontend_script` in the `register_block_type` server side, having the declarativness should allow us later to support lazy loading these scripts only when the block is used. 

cc @gziolo @aristath might have ideas on this PR too.

Reply:
What would be the difference with `script` that already exists and can be rendered on both the front-end and in the editor? How about CSS? Should it follow it?

Reply:
I'm not aware of it, could you point me to an example?

Reply:
> What would be the difference with script that already exists and can be rendered on both the front-end and in the editor? How about CSS?

General tradeoffs here are loading unnecessary JS/CSS on the frontend (we typically end up needing some specific editor only handling for a block) vs potentially having drifting implementations.

Reply:
By `script` I mean https://github.com/WordPress/gutenberg/blob/da8555d258edbe676fa079fb51252f918ae032e1/docs/designers-developers/developers/block-api/block-metadata.md#script.

Reply:
I see. The existing usages of `editorStyle` within Gutenberg are all pointing to the name of a registered style. I was able to get this working doing the same with `"script"`, in which case we'd still need to register the script somewhere.

I was not able to get this working using a relative file path following the instructions [here.](https://github.com/WordPress/gutenberg/blob/da8555d258edbe676fa079fb51252f918ae032e1/docs/designers-developers/developers/block-api/block-metadata.md#wpdefinedasset)

I tried the following two options, with `build/block-library-frontend/index.js` being the path of the JS file we want to have both in the editor and the frontend.

```js
// This assumes that the block.json file is at `build/block-library/blocks/<blockName>/block.json`
{
  "script": "file:../../../block-library-frontend/index.js"
}
```

```js
// This assumes that block.json is at `packages/block-library/src/<blockName>/block.json`
{
  "script": "file:../../../../build/block-library-frontend/index.js "
}
```

Even if those worked, using relative paths does not seem practical, given how hard it would be to determine where the target script is.

I've updated the PR with the usage of the `"script"` property that worked. It has the advantage of being clear about which block is responsible for pulling the JS, with the tradeoff being that it's an extra step.

Reply:
> The `WPDefinedAsset` type is a subtype of string, where the value represents a path to a JavaScript or CSS file relative to where `block.json` file is located.

The JS file should be next to `block.json` and CSS files in the `build` folder:

<img width="323" alt="Screen Shot 2021-03-05 at 07 14 42" src="https://user-images.githubusercontent.com/699132/110075077-8a090700-7d82-11eb-8940-f683929197db.png">

So it'd be:
```json
{
  "script": "file:./front.js "
}
```

> The existing usages of `editorStyle` within Gutenberg are all pointing to the name of a registered style.

It's only temporary because @aristath couldn't make it work with the file reference because some features were missing in the WordPress core. There might be more reasons, but we definitely want to use file references everywhere.

There is also an important consideration. The folder structure needs to be portable to the WordPress core as it is going to use the same `block.json` and other files.

Reply:
One more thing, in #22754, we added logic that automatically enqueues the file defined with `script` on the front-end when the block is rendered.

Reply:
> It's only temporary because @aristath couldn't make it work with the file reference because some features were missing in the WordPress core. There might be more reasons, but we definitely want to use file references everywhere.

I've added logic to webpack so that `script.js` gets built in the correct place, and tested loading it through block.json — it works fine, so I've removed the script registration for now. 

Pathing feel a bit hardcod-y, but it works fine as the only blocks we're interested in here are in the `block-library` package.

Reply:
> One more thing, in #22754, we added logic that automatically enqueues the file defined with `script` on the front-end when the block is rendered. It was also backported to the WordPress core:

The following comment indicates that the optimization made there does not keep assets loaded through block.json from being loaded everywhere for Core Blocks.

> > The assets associated with core block's are always enqueued ([source](https://github.com/WordPress/wordpress-develop/blob/64a2e8d55eba5b91bc7fff382676bd5873feaa38/src/wp-includes/script-loader.php#L2452-L2484)). Thus, they won't be able to take advantage of the optimizations here. At a minimum, the core blocks should be updated to reference the script and style asset handles themselves, rather than creating an ad hoc enqueuing implementation. This should serve as a simplifying refactoring, and ensure consistent behavior of assets enqueuing for all blocks.
> * A separate effort should seek to try to split the core blocks' script and style assets to per-block assets. In the meantime, there's not much benefit to these optimizations for core blocks.

I compared both approaches, and what the comment mentions checks out — when defining the script property for a Core Block, JS is loaded regardless of the blocks presence in the post.

As the changes in this PR are meant to affect only Core Blocks, it seems like the best course of action is still loading these files through the block's `render_callback`; this ensures that the JavaScript is loaded only if the block enqueuing it is present.

Reply:
> I compared both approaches, and what the comment mentions checks out — when defining the script property for a Core Block, JS is loaded regardless of the blocks presence in the post.

Do you mean that you confirmed with tests that the script defined with `script` in `block.json` enqueue on the front-end even when it's not rendered? I'm sure it worked properly at least for CSS with the changes that @aristath landed in #25220. Ah, I see now, it only applies to FSE themes or themes that opt-in for the behavior with `load_separate_block_styles`. By the way, we should change the name of the filter before backporting it to the WordPress core to cover the scripts part as well.